src/dune_engine/build.ml

```diff
@@ -353,79 +359,116 @@ let fold_labeled (type acc) t ~(init : acc) ~f =
+
+      let eval m = Execution.exec m.t
+    end
+
+    module Memo = Memo.Poly.Async (Function)
+
+    let exec t =
+      let open Fiber.O in
+      let file_exists = Fdecl.get file_exists_fdecl in
+      let eval_pred = Fdecl.get eval_pred_fdecl in
+      let rec go : type a. a t -> (a * Dep.Set.t) Fiber.t =
+        fun t ->
+          match t with
+          | Pure x -> Fiber.return (x, Dep.Set.empty)
+          | Map (f, a) ->
+            let+ a, dyn_deps_a = go a in
+            (f a, dyn_deps_a)
+          | Both (a, b) ->
+            let* a, dyn_deps_a = go a in
+            let+ b, dyn_deps_b = go b in
+            ((a, b), Dep.Set.union dyn_deps_a dyn_deps_b)
+          | Seq (a, b) ->
+            let* (), dyn_deps_a = go a in
+            let+ b, dyn_deps_b = go b in
+            (b, Dep.Set.union dyn_deps_a dyn_deps_b)
+          | Map2 (f, a, b) ->
+            let* a, dyn_deps_a = go a in
+            let+ b, dyn_deps_b = go b in
+            (f a b, Dep.Set.union dyn_deps_a dyn_deps_b)
+          | Deps _ -> Fiber.return ((), Dep.Set.empty)
+          | Paths_for_rule _ -> Fiber.return ((), Dep.Set.empty)
+          | Paths_glob g -> Fiber.return ((eval_pred g : Path.Set.t), Dep.Set.empty)
+          | Contents p -> Fiber.return (Io.read_file p, Dep.Set.empty)
+          | Lines_of p -> Fiber.return (Io.lines_of_file p, Dep.Set.empty)
+          | Dyn_paths t ->
+            let+ (x, paths), dyn_deps = go t in
+            (x, Dep.Set.add_paths dyn_deps paths)
+          | Dyn_deps t ->
+            let+ (x, dyn_deps), dyn_deps_x = go t in
+            (x, Dep.Set.union dyn_deps dyn_deps_x)
+          | Label _ -> Fiber.return ((), Dep.Set.empty)
+          | Fail { fail } -> fail ()
+          | If_file_exists (p, then_, else_) ->
+            if file_exists p then
+              go then_
+            else
+              go else_
+          | Filter_existing_files p ->
+            let+ (x, files), dyn_deps = go p in
+            let files = Path.Set.filter ~f:file_exists files in
+            ((x, files), dyn_deps)
+          | Catch (t, on_error) -> (
+              let+ res = Fiber.fold_errors
+                           ~init:None
+                           ~on_error:(fun bt _ -> Some (on_error bt.Exn_with_backtrace.exn))
+                           (fun () -> go t)
+              in
+              match res with
+              | Ok r -> r
+              | Error None -> assert false
+              | Error (Some r) -> (r, Dep.Set.empty)
+            )
+          | Memo m -> Memo.eval m
+          | Fiber f ->
+            let+ f = f in
+            (f, Dep.Set.empty)
+          | Dyn_fiber f ->
+            let* f, deps = go f in
+            let+ f = f in
+            (f, deps)
+          | Build b ->
+            let* b, deps0 = go b in
+            let+ r, deps1 = go_deps b in
+            (r, Dep.Set.union deps0 deps1)
+      and go_deps : type a. a t -> (a * Dep.Set.t) Fiber.t =
```

Comment:
Can this definition be inlined into the `Build` case? I think that would make the code a bit easier to read.

P.S.: Interesting, we seem to have a whole build system embedded here!